sympy/solvers/ode/single.py

```diff
@@ -0,0 +1,386 @@
+        self.eq = eq
+        self.eq_orig = eq_orig
+        self.func = func
+        self.sym = sym
+
+    # TODO: Add methods that can be used by many ODE solvers:
+    # order
+    # is_linear()
+    # get_linear_coefficients()
+    # eq_prepared (the ODE in prepared form)
+
+
+class SingleODESolver:
+    """
+    Base class for Single ODE solvers.
+
+    Subclasses should implement the _matches and _get_general_solution
+    methods. This class is not intended to be instantiated directly but its
+    subclasses are as part of dsolve.
+
+    Examples
+    ========
+
+    You can use a subclass of SingleODEProblem to solve a particular type of
+    ODE. We first define a particular ODE problem:
+
+    >>> from sympy import Eq, Function, Symbol
+    >>> x = Symbol('x')
+    >>> f = Function('f')
+    >>> eq = f(x).diff(x, 2)
+
+    Now we solve this problem using the NthAlgebraic solver which is a
+    subclass of SingleODESolver:
+
+    >>> from sympy.solvers.ode.single import NthAlgebraic, SingleODEProblem
+    >>> problem = SingleODEProblem(eq, f(x), x)
+    >>> solver = NthAlgebraic(problem)
+    >>> solver.get_general_solution()
+    [Eq(f(x), _C*x + _C)]
+
+    The normal way to solve an ODE is to use dsolve (which would use
+    NthAlgebraic and other solvers internally). When using dsolve a number of
+    other things are done such as evaluating integrals, simplifying the
+    solution and renumbering the constants:
+
+    >>> from sympy import dsolve
+    >>> dsolve(eq, hint='nth_algebraic')
+    Eq(f(x), C1 + C2*x)
+    """
+
+    # Subclasses should store the hint name (the argument to dsolve) in this
+    # attribute
+    hint = None  # type: ClassVar[str]
+
+    # Subclasses should define this to indicate if they support an _Integral
+    # hint.
+    has_integral = None  # type: ClassVar[bool]
+
+    # The ODE to be solved
+    ode_problem = None  # type: SingleODEProblem
+
+    # Cache whether or not the equation has matched the method
+    _matched = None  # type: Optional[bool]
+
+    def __init__(self, ode_problem):
+        self.ode_problem = ode_problem
+
+    def matches(self) -> bool:
+        if self._matched is None:
+            self._matched = self._matches()
+        return self._matched
+
+    def get_general_solution(self, *, simplify: bool = True) -> List[Equality]:
+        if not self.matches():
+            msg = "%s solver can not solve:\n%s"
+            raise ODEMatchError(msg % (self.hint, self.ode_problem.eq))
+        return self._get_general_solution()
+
+    def _matches(self) -> bool:
+        msg = "Subclasses of SingleODESolver should implement matches."
+        raise NotImplementedError(msg)
+
+    def _get_general_solution(self, *, simplify: bool = True) -> List[Equality]:
+        msg = "Subclasses of SingleODESolver should implement get_general_solution."
+        raise NotImplementedError(msg)
+
+    def get_numbered_constants(self, num=1, start=1, prefix='C'):
+        """
```

Comment:
The constants are supposed to be renamed by dsolve later so we can use any symbols we like while constructing the solution here. That means we don't need to allow specifying the prefix or anything else. I would suggest just having a single function that generates an iterator of numbered `Dummy` symbols like
```python
def iter_dummy_constants(num=None):
    if num is None:
        indices = itertools.count(1)
    else:
        indices = range(1, num+1)
    return (Dummy('C%d' % (n,)) for n in indices)
```

Reply:
note, too: `return numbered_symbols('C', cls=Dummy, start=1 if num is None else num)`

Reply:
> note, too: `return numbered_symbols('C', cls=Dummy, start=1 if num is None else num)`

Thanks, @smichr, and @oscarbenjamin for your suggestions.

Reply:
> The constants are supposed to be renamed by dsolve later so we can use any symbols we like while constructing the solution here. That means we don't need to allow specifying the prefix or anything else. I would suggest just having a single function that generates an iterator of numbered `Dummy` symbols like

@oscarbenjamin , I just found out that the problem with that is it sometimes produces symbols like `_C1` instead of `C1`. This causes a problem in `dsolve` when looking for `best_hint` here.

https://github.com/sympy/sympy/blob/573204b0feddfbbf9e5da0033bd52985088dc56e/sympy/solvers/ode/ode.py#L665-L672

Reply:
What problem does it cause?

Reply:
In this case, the problem is that `retdict["best"] `is `Eq(f(x), C1/x)` but the `FirstLinear` solution is `Eq(f(x), _C1/x)`. Now when `retdict["best"]` is compared to given hints, the equality does not match in line 670 which causes it to skip to another hint. The same happens when `C = Dummy()` is used.

Reply:
That sounds like something that should be fixed somewhere. For now though I guess we can just use `get_numbered_constants` as you did above.

Reply:
Should this PR or #18403  be merged?